Doc/library/os.rst

```diff
@@ -2292,6 +2292,8 @@ features:
    will fail with an :exc:`OSError` subclass in a number of cases:
 
    On Windows, if *dst* exists a :exc:`FileExistsError` is always raised.
+   The operation will fail if *src* and *dst* are on different filesystems. Use
```

Comment:
I'd say it "may" fail, and move this out of the "if dst exists" section (which is introduced in the preceding paragraph).

Reply:
Hi Zooba,

Thanks for the comment. The original (bpo-28356) talks about this, the main point of this documentation update is to alert users that it automatically will fail when moving across different disk drives (file systems).

This is something that is explained in errors in Python 3.8.10 but is missed in the documentation. See [my terminal](https://i.imgur.com/iwe1UQM.png)

As for the other point, I read it as being grouped together with Windows, as the next paragraph talks about Unix systems. I could rewrite the whole Windows paragraph, but I thought it better to tack it on to existing content.

I will leave it alone for now unless there is more discussion about this on either the bpo or in this thread.

Reply:
I am not a Windows dev, but @eryksun also said on the issue that it will always fail because of the specific Windows API call used in the implementation. On the other hand, perhaps we should say "may" so we can get away with changing the behavior in the future if we want.

As for the placement, the current docs (https://docs.python.org/3.10/library/os.html#os.rename) aren't very clear. The Unix paragraph has a bunch of other information beyond behavior if `dst` already exists. I'd actually suggest removing the "If dst exists, the operation will fail with an [OSError](https://docs.python.org/3.10/library/exceptions.html#OSError) subclass in a number of cases:" sentence completely. That way, we just have one paragraph talking about Windows oddities and one talking about Unix oddities.

Reply:
The implementation of `os.rename()` in Windows calls `MoveFileExW()` without the flag `MOVEFILE_COPY_ALLOWED`. The documentation states that "[i]f the destination is on another drive, you **must** [emphasis added] set the `MOVEFILE_COPY_ALLOWED` flag in `dwFlags`". This flag is not used by `os.rename()` or `os.replace()` because the destination is not supposed to be a copied file that has a new ID in the system.

POSIX [`rename()`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html) allows a system to support links across file systems, but the inode number -- and presumably the `(dev, ino)` tuple that uniquely identifies a file in the system -- must remain the same. I haven't seen this implemented. I've only ever seen rename fail with `EXDEV` in this case.

That said, using "may fail" is fine. A new version of NTFS and the Windows I/O manager could implement a type of reparse point that uses the [`FILE_ID_INFO`](https://docs.microsoft.com/en-us/windows/win32/api/winbase/ns-winbase-file_id_info), which in theory could be used to rename and hard link across two NTFS file systems.

Reply:
Yeah, there's a difference between "will always fail with the current implementation" and "will always fail for all time". This is the first case, but if we write it in the docs, we've made it the second case.

Better to leave ourselves open here, in case we (or Microsoft) decide to enable it later. Then we don't need to deprecate the function for two releases before changing it.